src/query/sql/src/planner/optimizer/rule/rewrite/rule_fold_constant.rs

```diff
@@ -0,0 +1,182 @@
+use crate::plans::RelOp;
+use crate::plans::RelOperator;
+use crate::plans::ScalarExpr;
+
+pub struct RuleFoldConstant {
+    id: RuleID,
+    patterns: Vec<SExpr>,
+    func_ctx: FunctionContext,
+}
+
+impl RuleFoldConstant {
+    pub fn new(func_ctx: FunctionContext) -> Self {
+        Self {
+            id: RuleID::FoldConstant,
+            patterns: vec![
+                // Filter
+                //  \
+                //   *
+                SExpr::create_unary(
+                    PatternPlan {
+                        plan_type: RelOp::Filter,
+                    }
+                    .into(),
+                    SExpr::create_leaf(
+                        PatternPlan {
+                            plan_type: RelOp::Pattern,
+                        }
+                        .into(),
+                    ),
+                ),
+                // EvalScalar
+                //  \
+                //   *
+                SExpr::create_unary(
+                    PatternPlan {
+                        plan_type: RelOp::EvalScalar,
+                    }
+                    .into(),
+                    SExpr::create_leaf(
+                        PatternPlan {
+                            plan_type: RelOp::Pattern,
+                        }
+                        .into(),
+                    ),
+                ),
+                // ProjectSet
+                //  \
+                //   *
+                SExpr::create_unary(
+                    PatternPlan {
+                        plan_type: RelOp::ProjectSet,
+                    }
+                    .into(),
+                    SExpr::create_leaf(
+                        PatternPlan {
+                            plan_type: RelOp::Pattern,
+                        }
+                        .into(),
+                    ),
+                ),
+                // Exchange
+                //  \
+                //   *
+                SExpr::create_unary(
+                    PatternPlan {
+                        plan_type: RelOp::Exchange,
+                    }
+                    .into(),
+                    SExpr::create_leaf(
+                        PatternPlan {
+                            plan_type: RelOp::Pattern,
+                        }
+                        .into(),
+                    ),
+                ),
+                // Join
+                //  \
+                //   *
+                SExpr::create_unary(
+                    PatternPlan {
+                        plan_type: RelOp::Join,
+                    }
+                    .into(),
+                    SExpr::create_leaf(
+                        PatternPlan {
+                            plan_type: RelOp::Pattern,
+                        }
+                        .into(),
+                    ),
+                ),
+            ],
+            func_ctx,
+        }
+    }
+
+    fn fold_constant(&self, scalar: &ScalarExpr) -> Result<ScalarExpr> {
```

Comment:
`as_expr` will turn agg into ColumnRef.
`from_expr` will not turn back into agg.